### PR TITLE
Redeem and unstake integration test

### LINKED ISSUE #611: Redeem and Unstake integration test (hashlock)

### PR DESCRIPTION
**Pull request by sarvesh-ost**
Fixes #611 
Redeem and un-stake integration tests.

### CODE REVIEW

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+            redeemer: auxiliaryAccounts[2],
+            bounty: await cogateway.bounty.call(),
+            nonce: await cogateway.getNonce.call(auxiliaryAccounts[2]),
+            beneficiary: originAccounts[2],
+            hashLock: hasher.l,
+            unlockSecret: hasher.s,
+        };
+
+        originAnchor = new Anchor(
+            auxiliaryWeb3,
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: `originAnchor` and `auxiliaryWeb3`? Does this pass right now? 😱

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        );
+        progressUnstakeAssertion = new ProgressUnstakeAssertion(
+            gateway,
+            token,
+            baseToken,
+        );
+        proofUtils = new ProofUtils(auxiliaryWeb3, originWeb3);
+    });
+
+    it('redeem', async () => {
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]:
```suggestion
    it('redeems', async () => {
```

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            event,
+            redeemRequest,
+            initialBalancesBeforeProgress,
+            facilitator,
+        );
+    });
+});
+
+/**
+ * This approve cogateway for redeem amount by unwraping base token.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]:
```suggestion
 * This approves the cogateway for redeem amount by unwrapping the base token.
```

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+ * @param {Object} ostPrime OSTPrime contract instance.
+ * @param {Object} redeemRequest Redeem request object.
+ * @param {Object }cogateway CoGateway contract instance.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]:
```suggestion
 * @param {Object} cogateway CoGateway contract instance.
```

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        cogateway.address,
+        redeemRequest.amount,
+        { from: redeemRequest.redeemer },
+    );
+}
+
+/**
+ * This returns gas consumption in a transaction.
+ * @param {Object} response Transaction object by truffle.
+ * @param {Web3 } web3
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]:
```suggestion
 * @param {Web3} web3
```

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+ * @return {Promise<BN>} gas consumption
+ */
+async function getTransactionGasConsumption(response, web3) {
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: As it returns consumed gas times gas price, the name is a little misleading.

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+     * @param {Web3} web3 Web3 instance.
+     */
+    constructor(cogateway, ostPrime, web3) {
+        this.cogateway = cogateway;
+        this.token = ostPrime;
+        this.web3 = web3;
+    }
+
+    /**
+     *  This verifies event and balances.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]:
```suggestion
     * This verifies event and balances.
```

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+     * @param {Object} event Event object after decoding.
+     * @param {RedeemRequest} redeemRequest Redeem request parameter.
+     * @param {number}gasConsumption Gas consumption in redeem request.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]:
```suggestion
     * @param {number} gasConsumption Gas consumption in redeem request.
```

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            redeemRequest,
+            initialBalances,
+            gasConsumption,
+        );
+
+        ProgressRedeemAssertion._assertProgressRedeemEvent(event, redeemRequest);
+    }
+
+    /**
+     * This capture base token and token balance of cogateway and redeemer
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]:
```suggestion
     * This captures base token and token balance of cogateway and redeemer.
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+        this.cogateway = cogateway;
+        this.token = ostPrime;
+        this.web3 = web3;
+    }
+
+    /**
+     *  This verifies event and balances.
+     * @param {Object} event Event object after decoding.
+     * @param {Object} event Event object after decoding.
+     * @param {number}gasConsumption Gas consumption in redeem request.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
     * @param {number} gasConsumption Gas consumption in redeem request.
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+            },
+        };
+    }
+
+    /**
+     * This asserts balances of redeemer and cogateway after Redeem.
+     * @param redeemRequest Redeem request parameters.
+     * @param {Balances} initialBalances Initial balance of redeemer and cogateway
+     *                                   generated by captureBalances method.
+     * @param {number}gasConsumption Gas consumption in redeem request.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
     * @param {number} gasConsumption Gas consumption in redeem request.
```

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 1 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: O.o What's wrong here. 
Anchor uses auxiliaryWeb3 to get block and anchor state root to originAnchor.

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 2 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 3 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 4 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 5 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 6 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: Updated to `getTransactionFee`
2bb0923

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js (hunk 7 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]: getTransactionFee

Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]: [2bb0923](https://github.com/OpenSTFoundation/mosaic-contracts/commit/2bb09232070ad16cf48ef3c5bd9ba0874b5c2323)

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js (hunk 8 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js (hunk 9 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js (hunk 11 above)
Comment (sarvesh-ost) [reply test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]: 2bb0923

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js (hunk 1 above)
Comment (schemar) [reply test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: Ah, of course. This is contracts land where we don't need web3 to access origin 😅

File: test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+            event,
+            redeemRequest,
+            initialBalancesBeforeProgress,
+            facilitator,
+        );
+    });
+});
+
+/**
+ * This approves the cogateway for redeem amount by unwrapping the base token.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/01_redeem_and_unstake.js]: wrapping or unwrapping? 🤔

File: test_integration/03_redeem_and_unstake/utils/confirm_redeem_intent_assertion.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+ * @property {string} messageHash Identifier for redeem and unstake process.
+ * @property {BN} blockHeight Height at which anchor state root is done.
+ */
+
+/**
+ *  Class to assert confirm redeem intent.
+ */
+class ConfirmRedeemIntentAssertion {
+    /**
+     *  This verifies event.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/confirm_redeem_intent_assertion.js]:
```suggestion
     * This verifies event.
```

File: test_integration/03_redeem_and_unstake/utils/confirm_redeem_intent_assertion.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+ * @property {BN} nonce Redeem nonce.
+ * @property {string} beneficiary Address of beneficiary on origin chain.
+ * @property {string} hashLock Hash Lock provided by the redeemer.
+ * @property {string} unlockSecret Unlock secret to unlock hash lock.
+ * @property {string} messageHash Identifier for redeem and unstake process.
+ * @property {BN} blockHeight Height at which anchor state root is done.
+ */
+
+/**
+ *  Class to assert confirm redeem intent.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/confirm_redeem_intent_assertion.js]:
```suggestion
 * Class to assert confirm redeem intent.
```

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        assert.strictEqual(
+            expectedRedeemerTokenBalance.eq(finalBalances.token.redeemer),
+            true,
+            `Redeemer token balance must be ${expectedRedeemerTokenBalance.toString(10)}`
+          + ` instead of ${finalBalances.token.redeemer.toString(10)}`,
+        );
+    }
+
+    /**
+     * This assert event after Redeem method.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]:
```suggestion
     * Assert event after Redeem method.
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+const assert = require('assert');
+
+/**
+ * Redeem Request object contains all the properties for redeem and unStake.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]: This *could* be defined as a class and also reused in the test.

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+ * @property balances.token.gateway ERC20 balance of gateway contract.
+ * @property balances.token.beneficiary ERC20 balance of beneficiary.
+ * @property balances.token.stakeVault ERC20 balance of stakeVault contract.
+ * @property balances.baseToken.gateway Base token(ETH) balance of gateway.
+ * @property balances.baseToken.beneficiary Base token(ETH) balance of beneficiary.
+ * @property balances.baseToken.stakeVault Base token(ETH) balance of stakeVault
+ */
+
+/**
+ *  Class to assert event and balances after progress un-stake
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]:
```suggestion
 * Class to assert event and balances after progress un-stake
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+     * @param {Object} baseToken Truffle baseToken instance.
+     */
+    constructor(gateway, token, baseToken) {
+        this.gateway = gateway;
+        this.token = token;
+        this.baseToken = baseToken;
+    }
+
+    /**
+     *  This verifies event and balances of beneficiary, gateway and stakeVault.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]:
```suggestion
     * This verifies event and balances of beneficiary, gateway and stakeVault.
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+     * @param {Object} event Event object after decoding.
+     * @param  {RedeemRequest} redeemRequest Redeem request parameters.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]:
```suggestion
     * @param {RedeemRequest} redeemRequest Redeem request parameters.
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+    async _assertBalancesForUnStake(redeemRequest, initialBalances, facilitator) {
+        const finalBalances = await this.captureBalances(
+            redeemRequest.beneficiary,
+            facilitator,
+        );
+
+        const reward = redeemRequest.gasPrice.mul(redeemRequest.gasLimit);
+        const unstakeAmount = redeemRequest.amount.sub(reward);
+
+        // Assert gateway balance
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]:
```suggestion
        // Assert gateway balance.
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+        assert.strictEqual(
+            expectedFacilitatorTokenBalance.eq(finalBalances.token.facilitator),
+            true,
+            `Facilitator token balance must be ${expectedFacilitatorTokenBalance} instead `
+          + `of ${finalBalances.token.facilitator}`,
+        );
+    }
+
+    /**
+     * This assert event after progress unstake method.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]:
```suggestion
     * Assert event after progress unstake method.
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+     * @param {Object} event Object representing unstake progressed event.
+     * @param {RedeemRequest} redeemRequest Redeem request parameters.
+     * @private
+     */
+    static __assertProgressUnStakeEvent(event, redeemRequest) {
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]: What is the meaning of *two* underscores? 🤔

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+ * BaseToken(ETH) and OSTPrime ERC20 balance of cogateway, redeemer.
+ * @typedef {Object} Balances
+ * @property balances.ostPrime.cogateway ERC20 balance of cogateway contract.
+ * @property balances.ostPrime.redeemer ERC20 balance of beneficiary.
+ * @property balances.baseToken.cogateway Base token(ETH) balance of cogateway.
+ * @property balances.baseToken.redeemer Base token(ETH) balance of redeemer.
+ */
+
+/**
+ *  Class to assert event and balances after Redeem.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
 * Class to assert event and balances after Redeem.
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+     * @param {Web3} web3 Web3 instance.
+     */
+    constructor(cogateway, ostPrime, web3) {
+        this.cogateway = cogateway;
+        this.token = ostPrime;
+        this.web3 = web3;
+    }
+
+    /**
+     *  This verifies event and balances.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
     * This verifies event and balances.
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+            redeemRequest,
+            initialBalances,
+            transactionFees,
+        );
+
+        RedeemAssertion._assertRedeemEvent(event, redeemRequest);
+    }
+
+    /**
+     * This capture base token and token balance of cogateway and redeemer
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
     * This captures base token and token balance of cogateway and redeemer
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+
+        // Assert Redeem amount is transferred to cogateway.
+        assert.strictEqual(
+            expectedCoGatewayTokenBalance.eq(finalBalances.token.cogateway),
+            true,
+            `CoGateway token balance must be ${expectedCoGatewayBaseTokenBalance.toString(10)}`
+           + ` instead of ${finalBalances.token.cogateway.toString(10)}`,
+        );
+
+        // Assert redeemer balance
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
        // Assert redeemer balance.
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+        assert.strictEqual(
+            expectedRedeemerTokenBalance.eq(finalBalances.token.redeemer),
+            true,
+            `Redeemer token balance must be ${expectedRedeemerTokenBalance.toString(10)}`
+           + ` instead of ${finalBalances.token.redeemer.toString(10)}`,
+        );
+    }
+
+    /**
+     * This assert event after Redeem method.
```
Comment (schemar) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]:
```suggestion
     * This asserts event after Redeem method.
```

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+    constructor(cogateway, ostPrime, web3) {
+        this.cogateway = cogateway;
+        this.token = ostPrime;
+        this.web3 = web3;
+    }
+
+    /**
+     * This verifies event and balances.
+     * @param {Object} event Event object after decoding.
+     * @param {Object} event Event object after decoding.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]: Events two time? `redeemRequest` is missing.

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+     * @param {number} transactionFees Transaction fees in redeem request.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]: For the bigger amount, the number may not handle. This should be BN?

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 30]
```diff
@@ -0,0 +1,209 @@
...
+            token: {
+                cogateway: await this.token.balanceOf(this.cogateway.address),
+                redeemer: await this.token.balanceOf(redeemer),
+            },
+        };
+    }
+
+    /**
+     * This asserts balances of redeemer and cogateway after Redeem.
+     * @param redeemRequest Redeem request parameters.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]: Should this be 
` @param {RedeemRequest} redeemRequest Redeem request parameters.`

File: test_integration/03_redeem_and_unstake/utils/redeem_assertion.js [hunk 31]
```diff
@@ -0,0 +1,209 @@
...
+     * @param {Balances} initialBalances Initial balance of redeemer and cogateway
+     *                                   generated by captureBalances method.
+     * @param {number} transactionFees Transaction fees in redeem request.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/redeem_assertion.js]: This should be number or BN ?
```suggestion
     * @param {BN} transactionFees Transaction fees in redeem request.
```

File: test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js [hunk 32]
```diff
@@ -0,0 +1,269 @@
...
+
+        // Assert gateway token balance must not change.
+        assert.strictEqual(
+            expectedGatewayTokenBalance.eq(finalBalances.token.gateway),
+            true,
+            `Gateway token balance must be ${expectedGatewayBaseTokenBalance.toString(10)}`
+           + ` instead of ${finalBalances.token.gateway.toString(10)}`,
+        );
+
+
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/progress_unstake_assertion.js]: remove one new line

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+            token: {
+                cogateway: await this.token.balanceOf(this.cogateway.address),
+                redeemer: await this.token.balanceOf(redeemer),
+            },
+        };
+    }
+
+    /**
+     * This asserts balances of redeemer and cogateway after progress Redeem.
+     * @param redeemRequest Redeem request parameters.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]:
```suggestion
     * @param {RedeemRequest} redeemRequest Redeem request parameters.
```

File: test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+     * @param {Balances} initialBalances Initial balance of redeemer and cogateway
+     *                                   generated by captureBalances method.
+     * @param {number} transactionFees Transaction fees in redeem request.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/progress_redeem_assertion.js]:
```suggestion
     * @param {BN} transactionFees Transaction fees in redeem request.
```

**Pull request merged**